Added input features

pablobaxter opened the pull request:
Fixed some bugs, added inputevent feature, and added activity manager class to broadcast intents into the device.

vidstige commented:
Thanks for your pr, it was rather long. I'll look into it shortly.

vidstige commented on src/se/vidstige/jadb/AdbFilterInputStream.java:
```diff
@@ -1,12 +1,21 @@
-    public AdbFilterInputStream(InputStream inputStream) {
-        super(inputStream);
+
+    private Transport mTransport;
```
What's the point of this? The thought here is that this stream class just works on a `InputStream` without know what kind of stream that is.

vidstige replied:
What's the point of this? The thought here is that this stream class just works on a `InputStream` without know what kind of stream that is.

vidstige replied:
Please revert this file before merge.

pablobaxter replied:
Reason why I added this is because the Transport class opens both an input and output stream, but in here, we just pass the input stream and close that only. I can still remove if I missed the handling the Transport elsewhere.

vidstige commented on src/se/vidstige/jadb/JadbDevice.java:
```diff
@@ -126,11 +126,29 @@ public void executeShell(OutputStream output, String command, String... args) th
+            i = stream.read();
+        } while(i != -1);
+        stream.close();
+    }
```
These function should go on a separate class that _uses_ `JadbDevice`. Perhaps inside a class called `Dumpsys` or something like that. `BatteryManager`?

vidstige replied:
Also the code is duplicated between the two functions above. Please move out to common function.

vidstige replied:
Also please use try-with-resource to also close stream in presence of exceptions. See more [here](https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html).

vidstige commented on src/se/vidstige/jadb/JadbDevice.java:
```diff
@@ -126,11 +126,29 @@ public void executeShell(OutputStream output, String command, String... args) th
+    public void resetUsbPlug() throws IOException, JadbException {
+        InputStream stream = executeShell("dumpsys", "battery", "reset");
+        int i;
+        do{
```
Single space betweet `do` and `{`

vidstige commented on src/se/vidstige/jadb/JadbDevice.java:
```diff
@@ -126,11 +126,29 @@ public void executeShell(OutputStream output, String command, String... args) th
+    public void simulateUsbUnplug() throws IOException, JadbException {
+        InputStream stream = executeShell("dumpsys", "battery", "unplug");
+        int i;
+        do{
```
Single space betweet `do` and `{`

vidstige commented on src/se/vidstige/jadb/Stream.java:
```diff
@@ -20,4 +20,11 @@ public static String readAll(InputStream input, Charset charset) throws IOExcept
+
+    public static void flushRead(InputStream is) throws IOException {
+        int v;
+        do{
```
Single space between `do` and `{`. Also code is duplicated in `JadbDevice.java`. Never copy-paste code.

vidstige commented on src/se/vidstige/jadb/Stream.java:
```diff
@@ -20,4 +20,11 @@ public static String readAll(InputStream input, Charset charset) throws IOExcept
+        int v;
+        do{
+            v = is.read();
+        } while(v >=0);
```
space after `>=` and `0`

vidstige commented on src/se/vidstige/jadb/managers/ActivityManager.java:
```diff
@@ -0,0 +1,41 @@
+ */
+public class ActivityManager {
+
+    private final JadbDevice device;
```
👍  for `final`

vidstige commented on src/se/vidstige/jadb/managers/ActivityManager.java:
```diff
@@ -0,0 +1,41 @@
+        int i;
+        do{
+            i = stream.read();
+        } while (i != -1);
```
Why are you reading the entire stream and then closing? Can we just close here, or do we need to read the entire stream? In any case, move it out as a function. Perhaps a `Stream` class with a static method `skipAll` or something.

vidstige commented on src/se/vidstige/jadb/managers/ActivityManager.java:
```diff
@@ -0,0 +1,41 @@
+        List<String> args = new ArrayList<>();
+        args.add("broadcast");
+        intent.generate(args);
+        if(permission != null){
```
whitespac.e

vidstige commented on src/se/vidstige/jadb/managers/ActivityManager.java:
```diff
@@ -0,0 +1,41 @@
+        do{
+            i = stream.read();
+        } while (i != -1);
+        stream.close();
```
Here, use the try-with-resource idiom.

vidstige commented on src/se/vidstige/jadb/managers/Component.java:
```diff
@@ -0,0 +1,36 @@
+public class Component {
+    
+    private Package mPackage;
+    private String mComponent;
```
in `jadb` there is no `m` prefix for members.

vidstige commented on src/se/vidstige/jadb/managers/Component.java:
```diff
@@ -0,0 +1,36 @@
+    private Package mPackage;
+    private String mComponent;
+    
+    public Component(String pkg, String component){
```
space before `{`

vidstige commented on src/se/vidstige/jadb/managers/Component.java:
```diff
@@ -0,0 +1,36 @@
+        this(new Package(pkg), component);
+    }
+    
+    public Component(Package pkg, String component){
```
Space before `{`

vidstige commented on src/se/vidstige/jadb/managers/Component.java:
```diff
@@ -0,0 +1,36 @@
+        mComponent = component;
+    }
+    
+    Package getPackage(){
```
Space before `{`

vidstige commented on src/se/vidstige/jadb/managers/Component.java:
```diff
@@ -0,0 +1,36 @@
+        return mPackage;
+    }
+    
+    String getComponent(){
```
Space before `{`

vidstige commented on src/se/vidstige/jadb/managers/Component.java:
```diff
@@ -0,0 +1,36 @@
+        return mComponent;
+    }
+
+    String generateComponent(){
```
Space before `{`

vidstige commented on src/se/vidstige/jadb/managers/Component.java:
```diff
@@ -0,0 +1,36 @@
+    }
+    
+    @Override
+    public String toString(){
```
Space before `{`

vidstige commented on src/se/vidstige/jadb/managers/Intent.java:
```diff
@@ -0,0 +1,761 @@
+            flags.add("--receiver-replace-pending");
+        }
+    }
+}
```
In entire file:
* Fix whitespace
* remove `m` prefix for members.

vidstige commented on src/se/vidstige/jadb/managers/KeyEvent.java:
```diff
@@ -0,0 +1,1383 @@
+            System.out.println(line);
+        }
+    }
+}
```
In entire file:
* Why are some warning suppressed?
* Fix whitespace
* Let's keep all the constants in this file and move out the methods. Perhaps to `KeyEvent`?

vidstige commented on src/se/vidstige/jadb/managers/Package.java:
```diff
@@ -14,7 +14,9 @@ public Package(String name) {
-    public boolean equals(Object o) { return name.equals(o); }
+    public boolean equals(Object o) {
+        return o instanceof Package && name.equals(((Package) o).name);
+    }
```
:+1:

vidstige commented on src/se/vidstige/jadb/managers/PackageManager.java:
```diff
@@ -49,24 +49,32 @@ private void verifyOperation(String operation, String target, String result) thr
         InputStream s = device.executeShell("rm", "-f", Bash.quote(file.getPath()));
-        Stream.readAll(s, Charset.forName("UTF-8"));
+        Stream.flushRead(s);
+        s.close();
```
Nice :+1:
To make it even better, use try-with-resource idiom.

vidstige commented on src/se/vidstige/jadb/managers/PackageManager.java:
```diff
@@ -49,24 +49,32 @@ private void verifyOperation(String operation, String target, String result) thr
+        verifyOperation("install", apkName, result);
+        is.close();
+        s.close();
+    }
```
Revert changes to this method
* The apkFile is of type `File` to make it easier to understand what to pass, and also to avoid misstake. Don't chase bugs, the compiler can find! ;-)

vidstige commented on src/se/vidstige/jadb/managers/PackageManager.java:
```diff
@@ -49,24 +49,32 @@ private void verifyOperation(String operation, String target, String result) thr
+    }
+
+    public void install(String apkFileName, InputStream is) throws IOException, JadbException {
+        install(apkFileName, is, new ArrayList<String>(0));
```
Always use `File` for local file arguments.

vidstige commented on src/se/vidstige/jadb/managers/PackageManager.java:
```diff
@@ -86,10 +94,12 @@ public void uninstall(Package name) throws IOException, JadbException {
 
     public void launch(Package name) throws IOException, JadbException {
         InputStream s = device.executeShell("monkey", "-p", name.toString(), "-c", "android.intent.category.LAUNCHER", "1");
+        Stream.flushRead(s);
```
Why is this needed?

vidstige commented on test/se/vidstige/jadb/test/integration/RealDeviceTestCases.java:
```diff
@@ -114,4 +115,23 @@ public void testScreenshot() throws Exception {
+    public void testUsbReset() throws Exception{
+        JadbDevice any = jadb.getAnyDevice();
+        any.resetUsbPlug();
+    }
```
:+1: for adding tests.

vidstige commented on test/se/vidstige/jadb/test/integration/RealDeviceTestCases.java:
```diff
@@ -114,4 +115,23 @@ public void testScreenshot() throws Exception {
+        JadbDevice any = jadb.getAnyDevice();
+        KeyEvent.Builder keyEventBuilder = new KeyEvent.Builder(any);
+        keyEventBuilder.addKey(KeyEvent.KEYCODE_POWER).build().execute();
+    }
```
Would it not be easier if the interface was something like

```java
KeyManager keys = new KeyManager(any);
keys.press(Keys.KEYCODE_POWER);
```

pablobaxter commented:
Fixed the whitespaces in code. 

The reason for the flush is because closing the stream immediately after running `executeShell` seems to prevent the shell from running. Flushing the inputstream returned seemed to fix this. Added the `flushStream` function as a static method in the `Stream` class.

Added try-with-resource for as requested.

The original though with the keyinput interface was that you can simulate pressing multiple keys. I added the option to input multiple key codes using the manager interface you mentioned, and made the `longPress()` option into an actual key event that performs a long press on passed key codes.

Added BatteryManager class to handle usb connect/disconnect.

Reverted the apk install functions.

pablobaxter commented on src/se/vidstige/jadb/managers/Intent.java:
```diff
@@ -0,0 +1,761 @@
+            flags.add("--receiver-replace-pending");
+        }
+    }
+}
```
The `TODO` here are in case the documentation should be kept. All documentation is from the source file.

vidstige replied:
Did you write this file? Is it copy-pasted n whole or part? From where?

Remember to 
* fix whitespace
* Remove m prefix for memebers

However, if this is copy-pasted in whole, I will consider keeping a different coding style for this file.

pablobaxter commented on src/se/vidstige/jadb/managers/Intent.java:
```diff
@@ -0,0 +1,761 @@
+                StringBuilder builder = new StringBuilder(strings[0]);
+                for(int i = 1; i < strings.length; i++) {
+                    //TODO escape commas
+                    builder.append(",").append(strings[i]);
```
Thinking about it now, not sure if commas really need to be escaped.

vidstige commented:
Thanks for the cleanup!

Regarding the API, it makes a lot of sense to be able to send a combination of keys and an optional long press flag. But it would be nice if the most common case - Single key is also very easy. How about we have two functions like so

```java
KeyManager keys = new KeyManager(any);
keys.press(Keys.KEYCODE_POWER); // Single key press
keys.press(Keys.combination(Keys.A, Keys.B).longpress()); // Multiple keys and long press
```

Or something like that? Then the combination would be static in `Keys` class  and return a builder. The builder class would then need to be non-static to be able to access the `execute` (should probably be renamed) function in `KeyManager`.

pablobaxter commented:
The reason for the `execute` command was due to the input command allowing mouse, gamepad, trackball, etc. commands to be used. The idea was to be able to create an abstracted `InputManager` object that was prebuilt using a child class (`KeyManager`), and passing it around as needed to perform a the defined actions.

However, thinking about it further, I can create an `InputEvent` class that handles the `execute`, with an `InputManager` class creating each event, for example:

    InputEvent powerButton = InputManager.createKeyEvent(Keys.KEYCODE_POWER);
    InputEvent powerButtonLongPress = InputManager.createKeyEvent(Keys.KEYCODE_POWER, true);
    InputEvent mouseLeftClick = InputManager.createMouseEvent(Mouse.LEFT_CLICK);

This way, the input events can be passed and changed in one location vs many if the same event is called multiple times (power button for example) or passed into a function that batches them.

vidstige commented:
Yeah, alright. I guess my main concern is I would like some class to act a bit like the `PropertManager`, `ActivityManager` or `PackageManager` already in jadb, whose job is to know how to execute a binary on the device. I would not like this logic to be inside some small class representing what key to press. Does this make sense? E.g. you can build up your events and then pass that to the `EventManager` (name just an example) that would know how to execute commands on the device.

Secondly I don't think the interface has to be too generic, it will just make it hard to use. I think we can have both things though, so that there can be a simple `press` method for convenience that would just build an event under the hood.

pablobaxter commented:
So I updated the input event handling as follows:

        InputManager inputManager = new InputManager(any);
        inputManager.sendEvent(KeyEvent.press(KeyCode.KEYCODE_POWER));

The manager takes a type of `InputEvent`, with `KeyEvent` as a subclass, in case someone decides to write a `SwipeEvent` class or `TapEvent` class to handle these other input types.

vidstige commented on src/se/vidstige/jadb/AdbFilterInputStream.java:
```diff
@@ -44,4 +53,10 @@ public int read(byte[] buffer, int offset, int length) throws IOException {
+    public void close() throws IOException {
+        super.close();
+        transport.close();
+    }
```
Remember to revert this file, it should know about the transport class.

vidstige commented on src/se/vidstige/jadb/JadbDevice.java:
```diff
@@ -126,7 +126,7 @@ public void executeShell(OutputStream output, String command, String... args) th
     }
 
-    private int getMode(File file) {
+    public int getMode(File file) {
```
This does not need to be public, let's keep private

vidstige commented on src/se/vidstige/jadb/RemoteFile.java:
```diff
@@ -13,7 +13,7 @@
     public boolean isDirectory() { throw new UnsupportedOperationException(); }
 
-    public String getPath() { return path;}
+    public String getPath() { return path; }
```
:+1:

vidstige commented on src/se/vidstige/jadb/Stream.java:
```diff
@@ -20,4 +20,11 @@ public static String readAll(InputStream input, Charset charset) throws IOExcept
+        int v;
+        do {
+            v = is.read();
+        } while(v >= 0);
```
space before parenthesis (`(`)

vidstige commented on src/se/vidstige/jadb/managers/ActivityManager.java:
```diff
@@ -0,0 +1,40 @@
+        List<String> args = new ArrayList<>();
+        args.add("broadcast");
+        intent.generate(args);
+        if(permission != null) {
```
space after `if`.

vidstige commented on src/se/vidstige/jadb/managers/BatteryManager.java:
```diff
@@ -0,0 +1,35 @@
+    public void simulateUsbUnplug() throws IOException, JadbException {
+        try (InputStream stream = device.executeShell("dumpsys", "battery", "unplug")) {
+            Stream.flushRead(stream);
+            stream.close();
```
No, need for close here as you use try-with-resource?

vidstige commented on src/se/vidstige/jadb/managers/BatteryManager.java:
```diff
@@ -0,0 +1,35 @@
+    public void resetUsbPlug() throws IOException, JadbException {
+        try (InputStream stream = device.executeShell("dumpsys", "battery", "reset")) {
+            Stream.flushRead(stream);
+            stream.close();
```
No, need for close here as you use try-with-resource?

vidstige commented on src/se/vidstige/jadb/managers/InputManager.java:
```diff
@@ -0,0 +1,28 @@
+            throw new JadbException(e.getMessage());
+        }
+    }
+}
```
Clean class! :)

vidstige commented on src/se/vidstige/jadb/managers/InputManager.java:
```diff
@@ -0,0 +1,28 @@
+        try (InputStream is = device.executeShell("input", event.buildArgs())) {
+            Stream.flushRead(is);
+        } catch (IOException e) {
+            throw new JadbException(e.getMessage());
```
Here the important inner exception gets lost, please pass e to `JadbException` constructor.

vidstige commented on src/se/vidstige/jadb/managers/KeyCode.java:
```diff
@@ -0,0 +1,1328 @@
+ * limitations under the License.
+ *
+ * Author: pablobaxter.
+ */
```
IS this file copied in whole or part? Please remove the author tag here.

pablobaxter replied:
This file was copied mostly from the Android Intent source, but with custom changes to be able to bring it into this library. I'll still remove the author tag, and make any changes needed to fit the code style of the library.

vidstige replied:
Alright, great. Can you also provide a link with the original source at the top of the file to make it easier to update, etc.?

vidstige commented on src/se/vidstige/jadb/managers/KeyCode.java:
```diff
@@ -0,0 +1,1328 @@
+    public String toString() {
+        return String.valueOf(val);
+    }
+}
```
newline at end of file

vidstige commented on src/se/vidstige/jadb/managers/KeyEvent.java:
```diff
@@ -0,0 +1,36 @@
+    protected String[] buildArgs() {
+        ArrayList<String> args = new ArrayList<>();
+        args.add("keyevent");
+        if(longPress) {
```
space after `if`

vidstige commented on src/se/vidstige/jadb/managers/KeyEvent.java:
```diff
@@ -0,0 +1,36 @@
+    public static KeyEvent press(KeyCode key) {
+        return new KeyEvent(key);
+    }
+}
```
newline at end of file

vidstige commented on src/se/vidstige/jadb/managers/PackageManager.java:
```diff
@@ -48,8 +48,10 @@ private void verifyOperation(String operation, String target, String result) thr
-        Stream.readAll(s, Charset.forName("UTF-8"));
+        try (InputStream stream = device.executeShell("rm", "-f", Bash.quote(file.getPath()))) {
+            Stream.flushRead(stream);
+            stream.close();
```
Close should not be needed here as you use try-with-resource? What's wring with the `readAll` method? I feel the name is clearer.

pablobaxter replied:
Honestly, I just wanted to be able to close the stream, since the output does not matter for these cases. However, I found that closing the stream immediately prevented the command from being completed. In regards to using `flushRead` vs `readAll`, it was more of me being anal about seeing new objects created for no reason.

vidstige commented on src/se/vidstige/jadb/managers/PackageManager.java:
```diff
@@ -86,11 +90,14 @@ public void uninstall(Package name) throws IOException, JadbException {
-        s.close();
+        try (InputStream stream = device.executeShell("monkey", "-p", name.toString(), "-c", "android.intent.category.LAUNCHER", "1")) {
+            Stream.flushRead(stream);
+            stream.close();
```
Close should not be needed here as you use try-with-resource? What's wring with the `readAll` method? I feel the name is clearer.

pablobaxter commented:
I will hopefully have unit tests done this weekend.

vidstige commented:
Sweet, no stress. Just ping me here and I'll merge if everything looks good. Double check whitespace, variable names, etc in unit tests for smooth merge. :+1:

Thanks again for this awesome contribution.

vidstige commented:
Closing this for now as we also go merge conflict :-/
